src/test/isolation2/sql/resgroup/resgroup_views_seg_down.sql

```diff
@@ -0,0 +1,155 @@
+-- The test below checks that FTS can break the stuck cancelled/terminated
+-- query, if one of the segments has hanged.
+-- Original problem was:
+-- 1. For some reason (the reason itself is not important) at some point segment hangs.
+-- 2. Query to 'gp_resgroup_status_per_segment' also hangs.
+-- 3. If 'pg_cancel_backend' or 'pg_terminate_backend' is called before FTS makes
+-- promotion, the query stays is stuck condition.
+
+-- Change FTS settings to make FTS probe work faster in the test.
+!\retcode gpconfig -c gp_fts_probe_timeout -v 5 --masteronly;
+!\retcode gpconfig -c gp_fts_probe_retries -v 1 --masteronly;
+!\retcode gpstop -u;
+
+CREATE EXTENSION IF NOT EXISTS gp_inject_fault;
+
+include: helpers/server_helpers.sql;
+
+-- Case 1: check the scenario with 'pg_cancel_backend'.
+
+-- Reset FTS probe interval.
+SELECT gp_request_fts_probe_scan();
+
+-- Make a successfull query to allocate query executer backends on the segments.
+1:SELECT count(1) FROM gp_toolkit.gp_resgroup_status_per_segment WHERE rsgname='default_group';
+
+-- Emulate hanged segment condition:
+-- 1. Stop the backends from processing requests by injecting a fault.
+-- 2. Stop segment postmaster process. This can't be done by the fault injection,
+-- as we wouldn't be able to recover the postmaster back to life in this case.
+-- Thus do it by sending a STOP signal. We need to do it on all segments (not on
+-- the coordinator), as the segment process may be running on a separate machine.
+SELECT gp_inject_fault('exec_simple_query_start', 'suspend', dbid)
+FROM gp_segment_configuration WHERE role = 'p' AND content = 0;
+
+SELECT exec_cmd_on_segments('ps aux | grep ''dbfast1/demoDataDir0'' | awk ''FNR == 1 {print $2; exit}'' | xargs kill -STOP');
```

Comment:
why not
```suggestion
SELECT pg_ctl(datadir, 'stop', 'immediate') FROM gp_segment_configuration WHERE role = 'p' and content = 0;

```

Reply:
During debug I've tried to use pg_ctl with stop. And the issue is not reproduced if the segment is fully stutdown. 
The segment's socket is closed in this case, so it doesn't get to the `poll`.

Reply:
how about to use datadir from gp_segment_configuration instead dbfast1/demoDataDir0?

Reply:
I agree that it would be a right thing to do. But! As we need to execute the command on the segments, the function `exec_cmd_on_segments` is marked as `execute on all segments`. And GPDB forbids to use 'FROM ...' clause in queries with such functions :(

Reply:
```sql
create external web table test(text text) execute 'echo $GP_SEG_DATADIR' format 'text';
CREATE EXTERNAL TABLE
select gp_execution_segment(), * from test;
 gp_execution_segment |                    text                    
----------------------+--------------------------------------------
                    1 | /home/gpadmin/.data/6/dbfast2/demoDataDir1
                    2 | /home/gpadmin/.data/6/dbfast3/demoDataDir2
                    0 | /home/gpadmin/.data/6/dbfast1/demoDataDir0
(3 rows)
```

Reply:
Using of $GP_SEG_DATADIR (or $GP_SEG_PORT and etc) will not allow us to resume the stopped process.
Because, after FTS promotion, the mirror will become the active segment, so $GP_SEG_DATADIR will change to '<some path>/dbfast_mirror1/demoDataDir0'.
So the process of original primary segment will stay in stopped state.

The only reliable (I mean not altering after FTS promotion) source of datadir I can currently think of is the `gp_segment_configuration`. But mixing it with the functions executed on the segments, as I've said, imposes limitations on the planning of such queries. If there is a way to overcome this limitations (I haven't found it yet), I guess it would be pretty cumbersome.

So I suggest to keep current approach, as it is quit simple.

Reply:
may we use \gset to save datadir from gp_segment_configuration and then use it instead explicit dbfast1/demoDataDir0?

Reply:
\gset is not working in the `isolation` tests. But I managed to use other mechanism available in the `isolation` tests for that purpose.

Reply:
> Because, after FTS promotion, the mirror will become the active segment, so $GP_SEG_DATADIR will change to '/dbfast_mirror1/demoDataDir0'.

how about
```sql
create external web table test(text text) execute 'echo $GP_SEG_DATADIR' on segment 0 format 'text';
CREATE EXTERNAL TABLE
select * from test;
                    text                    
--------------------------------------------
 /home/gpadmin/.data/6/dbfast1/demoDataDir0
(1 row)
```

Reply:
> how about

Hm, still do not see how can it help.
Before the FTS promotion it returns the datadir of the primary:
```
select * from test;
 text                                                     
----------------------------------------------------------
 /usr/re/share/ws30-ADBDEV-5517/data/dbfast1/demoDataDir0 
(1 row)
```
but after FTS promotion it return the datadir of the mirror (which is now acting as primary):
```
select * from test;
 text                                                            
-----------------------------------------------------------------
 /usr/re/share/ws30-ADBDEV-5517/data/dbfast_mirror1/demoDataDir0 
(1 row)
```

We need to resume the stopped process, and to identify it we need the original datadir name.

So the workable solution is to store the datadir name in the env variable before FTS promotion and then use it elsewhere. And to do it we do not need the external table - select from gp_segment_configuration works just fine.